data/mods/BombasticPerks/perkdata/enchanting_master/main_ench_on_hit_effect.json

```diff
@@ -0,0 +1,122 @@
+[
+  {
+    "type": "enchantment",
+    "id": "ench_enchanting_master",
+    "condition": "ALWAYS",
+    "hit_you_effect": [ { "id": "spell_enchanting_master_hit_you" } ]
+  },
+  {
+    "id": "spell_enchanting_master_hit_you",
+    "type": "SPELL",
+    "name": "effect of enchanting weapon",
+    "description": "effect when you attack others",
+    "valid_targets": [ "self", "ally", "hostile", "ground" ],
+    "effect": "effect_on_condition",
+    "effect_str": "EOC_enchanting_master_hit_you",
+    "spell_class": "perk_enchanting_master",
+    "flags": [ "SILENT", "NO_PROJECTILE", "NO_FAIL", "NO_HANDS", "NO_LEGS" ],
+    "shape": "blast",
+    "base_casting_time": 0,
+    "base_energy_cost": 0,
+    "min_range": 10,
+    "max_range": 10,
+    "energy_source": "MANA",
+    "difficulty": 0,
+    "max_level": 0,
+    "teachable": false
+  },
+  {
+    "id": "EOC_enchanting_master_hit_you",
+    "type": "effect_on_condition",
+    "effect": [
+      {
+        "run_eoc_with": "EOC_enchanting_master_ench_effect",
+        "variables": { "flag": "ITEM_ENCHANTMENT_BLAZING", "eoc": "EOC_enchanting_master_blazing" }
+      },
+      {
+        "run_eoc_with": "EOC_enchanting_master_ench_effect",
+        "variables": { "flag": "ITEM_ENCHANTMENT_CHILLING", "eoc": "EOC_enchanting_master_chilling" }
+      },
+      {
+        "run_eoc_with": "EOC_enchanting_master_ench_effect",
+        "variables": { "flag": "ITEM_ENCHANTMENT_SHOCKING", "eoc": "EOC_enchanting_master_shocking" }
+      },
+      {
+        "run_eoc_with": "EOC_enchanting_master_ench_effect",
+        "variables": { "flag": "ITEM_ENCHANTMENT_KINETIC", "eoc": "EOC_enchanting_master_kinetic" }
+      },
+      {
+        "run_eoc_with": "EOC_enchanting_master_ench_effect",
+        "variables": { "flag": "ITEM_ENCHANTMENT_PROJECTING", "eoc": "EOC_enchanting_master_projecting" }
+      },
+      {
+        "run_eoc_with": "EOC_enchanting_master_ench_effect",
+        "variables": { "flag": "ITEM_ENCHANTMENT_GRIM", "eoc": "EOC_enchanting_master_grim" }
+      },
+      {
+        "run_eoc_with": "EOC_enchanting_master_ench_effect",
+        "variables": { "flag": "ITEM_ENCHANTMENT_VAMPIRIC", "eoc": "EOC_enchanting_master_vampiric" }
+      }
+    ]
+  },
+  {
+    "id": "EOC_enchanting_master_ench_effect",
+    "type": "effect_on_condition",
+    "//": "In this eoc, u is the victim, npc is avatar.",
+    "//flag": "The required flag",
+    "//eoc": "The corresponding triggered eoc. In this eoc, u is avatar, npc is the victim since I thick this is more intuitive.",
+    "condition": {
+      "and": [ { "expects_vars": [ "flag", "eoc" ] }, { "npc_has_wielded_with_flag": { "context_val": "flag" } }, "u_exists" ]
+    },
+    "effect": {
+      "run_eoc_with": { "id": "_EOC_enchanting_master_ench_effect", "effect": { "run_eocs": { "context_val": "eoc" } } },
+      "alpha_talker": "npc",
+      "beta_talker": "u"
```

Comment:
This will end up just being confusing. Please stick to the convention of u = avatar if the avatar is involved.

Reply:
This is to make sure that avatar/caster is always the alpha talker for all of the following EOCs
(since most EOCs need avatar as the alpha talker but not the beta talker. These EOCs are not only called in this EOC chain but also in other places directly. )
eg. "EOC_enchanting_master_ench_effect" -> "EOC_enchanting_master_blazing" (u is avatar, npc is victim)-> "EOC_enchanting_master_item_ench_trigger_count_minus_one" ( u is avatar )
```
  {
    "id": "EOC_enchanting_master_item_ench_trigger_count_minus_one",
    "type": "effect_on_condition",
    "//": "The trigger count will minus 1. If it becomes 0, the enchantment on this item will be removed.",
    "effect": [
      {
        "u_run_inv_eocs": "random",
        "search_data": [ { "wielded_only": true } ],
        "true_eocs": "EOC_enchanting_master_trigger_count_minus_one"
      }
    ]
  }
```
"EOC_enchanting_master_item_ench_trigger_count_minus_one" is also called directly in "EOC_enchanting_master_corrupting_on_kill_monster"